src/Dscp/DB/DSL/Interpret/Sqlite3.hs

```diff
@@ -0,0 +1,188 @@
+
+import Database.SQLite.Simple hiding (query)
+import Database.SQLite.Simple.ToField (ToField(..))
+import Database.SQLite.Simple.FromField (FromField(..))
+
+import Text.InterpolatedString.Perl6
+
+import Dscp.Core.Types ( SignedSubmission(..)
+                       , Submission(..)
+                       , SubmissionType(..)
+                       , SubmissionWitness(..)
+                       , SubmissionSig
+                       , Assignment(..)
+                       , AssignmentType(..)
+                       , Address(..)
+                       , CourseId(..)
+                       , Grade(..)
+                       )
+import Dscp.Crypto (Hash, Signature, PublicKey)
+import Dscp.DB.DSL.Class
+import Dscp.DB.SQLite
+import Dscp.DB.SQLite.Class
+import Dscp.Educator.Txs (PrivateTx(..), PrivateTxId)
+
+-- TODO(kir): split into separate .Instances module.
+instance FromField Address
+instance FromField CourseId
+instance FromField Grade
+instance FromField SubmissionSig
+instance ToField   PrivateTxId
+
+instance
+--    v-- GHC says it cand "find" Monad in superclasses (wat).
+    ( Monad m
+    , MonadSQLiteDB m
+    )
+    => MonadSearchTxObj m
+  where
+    runTxQuery (SELECTTx WHERE (TxIdEq pid)) =
+        getPrivateTxFromId (error "get pk from keyring here") pid
+
+    -- TODO (kir): find where the 'Obj'ects live.
+    runObjQuery (SELECTObj WHERE (ObjHashEq hash)) =
+        return Nothing
+
+    runTxsQuery (SELECTTxs WHERE txFilter) =
+        getPrivateTxsByFilter (error "get pk from keyring here") txFilter
+
+getPrivateTxsByFilter
+    :: MonadSQLiteDB m
+    => PublicKey
+    -> TxsFilterExpr
+    -> m [PrivateTx]
+getPrivateTxsByFilter pk filterExpr = do
+    let
+      tables = requiredTables filterExpr
+
+      -- The [qc||] cannot occur inside its {}-interpolator :(
+      -- Expression [qc| {[qc||]} |] gives an error.
+      additionalJoins =
+        if Subject `Set.member` tables
+        then [qc|
+
+            left join Courses     on  Assignment.course_id       = Courses.id
+            left join Subjects    on    Subjects.course_id       = Courses.id
+
+             |]
+        else [qc||]
+
+    map (packPrivateTxQuery pk) <$> query
+        [qc|
+            select    Submissions.student_addr,
+                      Submissions.contents_hash,
+
+                      Assignments.course_id,
+                      Assignments.contents_hash,
+                      Assignments.desc,
+
+                      Submissions.signature,
+
+                      Transactions.grade,
+                      Transactions.time
+
+            from      Transactions
+
+            left join Submissions on             submission_hash = Submissions.hash
+            left join Assignments on Submissions.assignment_hash = Assignments.hash
+            {
+                additionalJoins
+            }
+            where {buildWhereStatement filterExpr}
+            ;
+        |]
+        ()
+
+buildWhereStatement = go
```

Comment:
Explicit type signatures for top-level bindings, please (actually don't know how this isn't a warning :thinking:).